finality-aleph/src/new_network/split.rs

```diff
@@ -0,0 +1,231 @@
+use crate::new_network::{ComponentNetwork, Data, ReceiverComponent, SendError, SenderComponent};
+use aleph_bft::Recipient;
+use codec::{Decode, Encode};
+use futures::channel::mpsc;
+use log::warn;
+use std::{marker::PhantomData, sync::Arc};
+use tokio::{
+    sync::Mutex,
+    time::{interval, Duration},
+};
+
+const MAXIMUM_RETRY_MS: u64 = 500;
+
+/// Used for routing data through split networks.
+#[derive(Clone, Encode, Decode)]
+pub enum Split<D1: Data, D2: Data> {
+    Left(D1),
+    Right(D2),
+}
+
+#[derive(Clone)]
+struct LeftSender<D1: Data, D2: Data, S: SenderComponent<Split<D1, D2>>> {
+    sender: S,
+    phantom: PhantomData<(D1, D2)>,
+}
+
+impl<D1: Data, D2: Data, S: SenderComponent<Split<D1, D2>>> SenderComponent<D1>
+    for LeftSender<D1, D2, S>
+{
+    fn send(&self, data: D1, recipient: Recipient) -> Result<(), SendError> {
+        self.sender.send(Split::Left(data), recipient)
+    }
+}
+
+#[derive(Clone)]
+struct RightSender<D1: Data, D2: Data, S: SenderComponent<Split<D1, D2>>> {
+    sender: S,
+    phantom: PhantomData<(D1, D2)>,
+}
+
+impl<D1: Data, D2: Data, S: SenderComponent<Split<D1, D2>>> SenderComponent<D2>
+    for RightSender<D1, D2, S>
+{
+    fn send(&self, data: D2, recipient: Recipient) -> Result<(), SendError> {
+        self.sender.send(Split::Right(data), recipient)
+    }
+}
+
+struct LeftReceiver<D1: Data, D2: Data, R: ReceiverComponent<Split<D1, D2>>> {
+    receiver: Arc<Mutex<R>>,
+    translated_receiver: mpsc::UnboundedReceiver<D1>,
+    left_sender: mpsc::UnboundedSender<D1>,
+    right_sender: mpsc::UnboundedSender<D2>,
+}
+
+struct RightReceiver<D1: Data, D2: Data, R: ReceiverComponent<Split<D1, D2>>> {
+    receiver: Arc<Mutex<R>>,
+    translated_receiver: mpsc::UnboundedReceiver<D2>,
+    left_sender: mpsc::UnboundedSender<D1>,
+    right_sender: mpsc::UnboundedSender<D2>,
+}
+
+async fn forward_or_wait<D1: Data, D2: Data, R: ReceiverComponent<Split<D1, D2>>>(
+    receiver: &Arc<Mutex<R>>,
+    left_sender: &mpsc::UnboundedSender<D1>,
+    right_sender: &mpsc::UnboundedSender<D2>,
+) {
+    match receiver.try_lock().ok() {
+        Some(mut receiver) => match receiver.next().await {
+            Some(Split::Left(data)) => {
+                if left_sender.unbounded_send(data).is_err() {
+                    warn!(target: "aleph-network", "Failed send despite controlling receiver, this shouldn't've happened.");
+                }
+            }
+            Some(Split::Right(data)) => {
+                if right_sender.unbounded_send(data).is_err() {
+                    warn!(target: "aleph-network", "Failed send despite controlling receiver, this shouldn't've happened.");
+                }
+            }
+            None => {
+                left_sender.close_channel();
+                right_sender.close_channel();
+            }
+        },
+        None => {
+            interval(Duration::from_millis(MAXIMUM_RETRY_MS))
```

Comment:
I may be mistaken but with the usage of the async mutex we wouldn't need this arm at all

Reply:
We do, `try_lock()` is not `async`, so if the lock failed this would just fall through every time and do essentially a busy wait at best.

Reply:
I meant the usage of the https://docs.rs/async-mutex/latest/async_mutex/ (with `mutex.lock().await` instead of the try_lock)

Reply:
You are right! <3